### PR TITLE
Fixing Public Jira Issue : APIMANAGER-3794

### PR DESCRIPTION
**Pull request by ChamNDeSilva**
Adding thumbnail images when adding or updating apis through publisher rest api

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 1]
```diff
@@ -1895,6 +1956,17 @@ private static void checkFileSize(FileHostObject fileHostObject)
         }
     }
 
+    private static void checkImageSize(File file)
+            throws ScriptException, APIManagementException, IOException {
+
+        if(file.exists()){
+            long length = file.length();
+            if(length/ 1024.0 > 1024){
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Better use a constant for the max allowed file size.

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 2]
```diff
@@ -1712,6 +1745,34 @@ public static boolean jsFunction_updateAPI(Context cx, Scriptable thisObj,
...
+                    String thumbPath = APIUtil.getIconPath(api.getId());
+                    String thumbnailUrl = apiProvider.addIcon(thumbPath, thumbIcon);
+                    api.setThumbnailUrl(APIUtil.prependTenantPrefix(thumbnailUrl, api.getId().getProviderName()));
+
+                    /*Set permissions to anonymous role for thumbPath*/
+                    APIUtil.setResourcePermissions(api.getId().getProviderName(), null, null, thumbPath);
+
+                } catch (IOException e) {
+                    handleException("[Error] Cannot read data from the URL", e);
+                    return false;
```
Comment (nuwand) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: This line of code is unreachable. handleException will throw an exception out.

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java (hunk 1 above)
Comment (ChamNDeSilva) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Fixed in commit d7ce744e30264fb8028ee397dc3ed745fcc70142

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java (hunk 2 above)
Comment (ChamNDeSilva) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Fixed in commit d7ce744e30264fb8028ee397dc3ed745fcc70142

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 3]
```diff
@@ -1302,7 +1306,35 @@ public static boolean jsFunction_addAPI(Context cx, Scriptable thisObj,
...
+
+                Icon thumbIcon = new Icon(fileBody.getInputStream(), url.openConnection().getContentType());
+                String thumbPath = APIUtil.getIconPath(api.getId());
+                String thumbnailUrl = apiProvider.addIcon(thumbPath, thumbIcon);
+                api.setThumbnailUrl(APIUtil.prependTenantPrefix(thumbnailUrl, api.getId().getProviderName()));
+
+                    /*Set permissions to anonymous role for thumbPath*/
+                APIUtil.setResourcePermissions(api.getId().getProviderName(), null, null, thumbPath);
+
+            }catch (IOException e){
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Slight formatting issue space needed after }

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 4]
```diff
@@ -1390,6 +1422,7 @@ public static boolean jsFunction_updateAPI(Context cx, Scriptable thisObj,
         String bizOwner = (String) apiData.get("bizOwner", apiData);
         String bizOwnerEmail = (String) apiData.get("bizOwnerEmail", apiData);
         String visibility = (String) apiData.get("visibility", apiData);
+        String thumbUrl = (String) apiData.get("thumbUrl",apiData);
```
Comment (madusankapremaratne) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 5]
```diff
@@ -1712,6 +1745,34 @@ public static boolean jsFunction_updateAPI(Context cx, Scriptable thisObj,
                 // retain the previously uploaded image
                 api.setThumbnailUrl(oldApi.getThumbnailUrl());
             }
+
+            if(thumbUrl != null){
+                try {
+                    URL url = new URL(thumbUrl);
```
Comment (dineshjweerakkody) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: I noticed that we have this icon handling part in two places. Can we make it a one method and call in both places?

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 6]
```diff
@@ -1302,7 +1306,35 @@ public static boolean jsFunction_addAPI(Context cx, Scriptable thisObj,
         		PrivilegedCarbonContext.endTenantFlow();
         	}
         }
-        
+        if(thumbUrl != null){
+            try {
+                URL url = new URL(thumbUrl);
+                String imageType = url.openConnection().getContentType();
+
```
Comment (malakasilva) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: No need to keep lines and better have String in constants

Comment (ChamNDeSilva) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Fixed in https://github.com/wso2/carbon-apimgt/pull/1097

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 7]
```diff
@@ -1302,7 +1306,35 @@ public static boolean jsFunction_addAPI(Context cx, Scriptable thisObj,
         		PrivilegedCarbonContext.endTenantFlow();
         	}
         }
-        
+        if(thumbUrl != null){
```
Comment (prabathabey) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Formatting issues.

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java [hunk 8]
```diff
@@ -1895,6 +1956,17 @@ private static void checkFileSize(FileHostObject fileHostObject)
         }
     }
 
+    private static void checkImageSize(File file)
+            throws ScriptException, APIManagementException, IOException {
+
+        if(file.exists()){
```
Comment (prabathabey) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: More formatting issues.

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java (hunk 1 above)
Comment (prabathabey) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIProviderHostObject.java]: Some more formatting issues.

**Pull request merged**